src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,201 @@
+
+    starting_address = FermiFS2C(near_uniform(FermiFS{n_alpha_elec,n_orb}), near_uniform(FermiFS{n_beta_elec,n_orb}))
+
+    val_type = typeof(fd.int0)
+    addr_type = typeof(starting_address)
+    fd_type = typeof(fd)
+    MolecularHamiltonian{val_type,addr_type,fd_type}(starting_address, fd)
+end
+
+struct Modes{T<:FermiFSIndex}
+    occupied::Vector{Vector{T}}
+    unoccupied::Vector{Vector{T}}
+end
+
+function modes_extract(addr::FermiFS2C)
+    occupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    unoccupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    for (i, state) in enumerate(addr.components)
+        occupied_modes[i] = OccupiedModeMap(state)
+        state_unoccupied_modes = FermiFSIndex[]
+        for j = 1:num_modes(state)
+            ind = find_mode(state, j)
+            if ind.occnum == 0
+                push!(state_unoccupied_modes, ind)
+            end
+        end
+        unoccupied_modes[i] = state_unoccupied_modes
+    end
+    Modes(occupied_modes, unoccupied_modes)
+end
+
+struct MolecularHamiltonianOperatorColumn{A<:FermiFS2C,T,O<:MolecularHamiltonian,OD,M<:Modes{FermiFSIndex}} <: AbstractOperatorColumn{A,T,O}
+    addr::A # Contains address Psi_i, provided by operator_column
+    op::O   # Represent Hamiltonian itself, provided by operator_column
+    diag::T # < Psi_i | Psi_i >, calculated by diagonal_element
+    ods::OD # Vector [ < Psi_j | Psi_i > ], calculated by offdiagonals
+    modes::M # Store the modes of addr
+end
+
+function starting_address(h::MolecularHamiltonian)
+    return h.starting_address
+end
+
+function operator_column(h::MolecularHamiltonian{T,A,D}, a::FermiFS2C)::MolecularHamiltonianOperatorColumn where {T<:Number,A,D}
+    diag = zero(T)
+    ods = Vector{Float64}[]
+
+    modes = modes_extract(a)
+
+    one_elec_int = one_electron_integral(h.fcidump.int1, modes.occupied)
+    two_elec_int = two_electron_integral(h.fcidump.int2, modes.occupied)
+    diag = h.fcidump.int0 + one_elec_int + two_elec_int
+
+    MolecularHamiltonianOperatorColumn{typeof(a),T,typeof(h),typeof(ods),typeof(modes)}(a, h, diag, ods, modes)
+end
+
+function diagonal_element(column::MolecularHamiltonianOperatorColumn{A,T,O,OD}) where {A<:FermiFS2C,T,O,OD}
+    return column.diag
+end
+
+function num_offdiagonals(column::MolecularHamiltonianOperatorColumn)
+    n_spin_orb = headvar(column.op.fcidump, "NORB") * 2
+    n_elec = headvar(column.op.fcidump, "NELEC")
+
+    # One-electron excitation
+    n_one_electron_excitation = n_elec * (n_spin_orb - n_elec)
+
+    # Two-electron excitation
+    n_two_electron_excitation = binomial(n_elec, 2) * binomial(n_spin_orb - n_elec, 2)
+
+    n_one_electron_excitation + n_two_electron_excitation
+end
+
+const ModeIndex = @NamedTuple{ch::Int, orb::Int}
+
+struct ModeTransition
+    old::ModeIndex
+    new::ModeIndex
+end
+
+mutable struct MolecularHamiltonianOffDiagonalsIterState
+    occ_ind::Int
+    unocc_ind::Int
+    spin_ch::Int
+end
```

Comment:
I think @mtsch provided an example for an iterator without a mutable state. The problem with `mutable struct` is that they cause heap allocations as well.